[Feat] Document rules engine (#86)
libremfg/libremfg.github.io: content/how-to/publish-subscribe/turn-value-into-event.md

---
title: 'Turn values into events'
date: '2024-03-05T19:10:43-03:00'
draft: true
categories: ["how-to"]
description: Use the Rhize rules engine to turn a value from a data source into an event for orchestration
weight:
menu:
  main:
    parent: howto-pubsub
    identifier:
---

> A {{< abbr "data source" >}} might emit a high volume of data,
> but not all this data is useful.

Author: ```suggestion
A {{< abbr "data source" >}}  can emit a high volume of data,
but not all this data is useful.
```

To filter the data for significant events, use the _rules engine_ to write rules to trigger workflows.

The Rhize rules engine evaluates incoming values from a data source and then converts significant changes into _events_ to be processed by a {{< abbr "bpmn" >}} workflow.
Each rule exists on the level of the {{< abbr "equipment class" ->}}.

> Rules work by evaluating the selected properties for a specified change.

Author: ```suggestion
Each rule evaluates changes that happen to a set of specified properties for a data source.
```

> If the evaluation condition is true, the engine throws a trigger to start a workflow.

Author: ```suggestion
If the condition evaluates to `true`, the engine throws a trigger to start a workflow.
```

You can also use the message payload to create variables for the workflow to use.

```mermaid
---
title: Rules trigger workflows from data-source changes
---
flowchart LR
    A(Property\nchanged?) -->|yes| B{"rule evaluates\nto true?"}
    B -->|no| C(do nothing)
    B -->|"yes\n(optional: pass variables)"| D(Run BPMN workflow)
```

## Prerequisites

> To create a rule, you first must create the models and workflows that you want to associate with this rule.

Author: ```suggestion
To create a rule, you must first create the models and workflows that you want to associate with this rule.
```

The procedure is as follows:

> 1. [Create a data source]({{< ref "connect-datasource" >}})

Author: ```suggestion
1. [Create a data source]({{< ref "connect-datasource" >}}).
```

1. [Write a BPMN workflow]({{< relref "/how-to/bpmn/create-workflow/" >}}) that the rule triggers. You can edit or change this workflow later.

> 1. [Create models]({{< relref "../model" >}}) for the following associated objects

Author: ```suggestion
1. [Create models]({{< relref "../model" >}}) for the following associated objects:
```

    - A [{{< abbr "data source" >}}]({{< ref "master-definitions#data-source" >}})
    - An [{{< abbr "equipment" >}}]({{< ref "master-definitions#equipment" >}}) item bound to this data source
    - An [{{< abbr "equipment class" >}}]({{< ref "master-definitions#equipment-class" >}}) that this equipment belongs to

Once you you've created these objects, you can create a rule for the equipment class, as documented in the following section.

## Create a rule in the UI

In the Rhize UI, select the equipment class that you want to create a rule for.
Then follow these steps:

1. In the Equipment class, select the **Rules** tab.
1. In **Workflow specification**, select the BPMN workflow that will orchestrate processes based on the event.

> 1. In **Trigger properties**, add the properties that the rule evaluates. If any of these properties change, the rule will evaluate them.

Author: ```suggestion
1. In **Trigger properties**, add the properties that the rule evaluates to. If any of these properties change, the rule will evaluate them.
```
